### PR TITLE
feat(website): tuples as json (read-only) and render each input for tuples in write mode

### PR DESCRIPTION
**Pull request by Ignusmart**
The primary goal of this pull request is to enable the rendering of tuples as JSON when they are in read-only mode, and to facilitate the rendering of individual inputs for tuples in write mode.

# Deploy View (Read Only)

## Before the adjustment

image

## With the adjustment

image

## Before the adjustment

image

## With the adjustment

image

# NOTES

Before merging, we should thoroughly test the new tuple functionality, particularly in write mode, especially in scenarios where functions expect nested tuples.

### CODE REVIEW

File: packages/website/src/features/Packages/FunctionInput.tsx [hunk 1]
```diff
@@ -31,10 +31,6 @@ export const FunctionInput: FC<{
     setDataArray(dataArray.splice(index, 1));
   };
 
-  useEffect(() => {
-    valueUpdated(dataArray.map((item) => item.val));
-  }, [dataArray]);
-
```
Comment (Ignusmart) [inline packages/website/src/features/Packages/FunctionInput.tsx:37]: This one was causing that default values were not parsed as expected on first render of the function input, getting wrong params values (into an array), this before the fix:

Screenshot 2024-04-06 at 11 07 27 AM

With this removal, here the right params values

Screenshot 2024-04-06 at 11 08 07 AM

File: packages/website/src/features/Packages/FunctionInput.tsx [hunk 2]
```diff
@@ -33,8 +33,9 @@ export const FunctionInput: FC<{
   };
 
   useEffect(() => {
+    if (!isArray) return;
     valueUpdated(dataArray.map((item) => item.val));
-  }, [dataArray]);
+  }, [dataArray, isArray]);
```
Comment (Ignusmart) [inline packages/website/src/features/Packages/FunctionInput.tsx:38]: This one was causing that default values were not parsed as expected on first render of the function input, getting wrong params values (into an array), this before the fix:

Screenshot 2024-04-06 at 11 07 27 AM

With this update, here the right params values

Screenshot 2024-04-06 at 11 08 07 AM

File: packages/website/src/features/Packages/FunctionInput/TupleInput.tsx [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+import { FunctionInput } from '../FunctionInput';
+import { useState } from 'react';
+import { BigNumber } from 'ethers';
+import { AbiParameter } from 'viem';
+
+const TupleInput = ({
+  input,
+  handleUpdate,
+}: {
+  input: any;
```
Comment (mjlescano) [inline packages/website/src/features/Packages/FunctionInput/TupleInput.tsx:11]: Can this be typed as `input: AbiParameter;` ?

File: packages/website/src/features/Packages/FunctionInput/TupleInput.tsx [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+  handleUpdate: (value: any) => void;
```
Comment (mjlescano) [inline packages/website/src/features/Packages/FunctionInput/TupleInput.tsx:12]: Also here `value: AbiParameter`

File: packages/website/src/features/Deploy/DisplayedTransaction.tsx [hunk 5]
```diff
@@ -103,6 +104,47 @@ export function DisplayedTransaction(props: {
     return val.toString();
   }
 
+  function renderInput(type: string, val: string): React.ReactElement {
+    if (type === 'tuple') {
+      return (
+        <CopyBlock
```
[the remark this answers is not in the document]
Comment (noahlitvin) [reply packages/website/src/features/Deploy/DisplayedTransaction.tsx:110]: I think monaco is too "heavy" for this? Also don't like that we don't really have control over the scroll with it.

**Comment by noahlitvin** [comment]
LGTM! Here's one edge case to consider: https://usecannon.com/packages/synthetix-omnibus/latest/8453-andromeda/interact/system/CoreProxy/0x32C222A9A159782aFD7529c87FA34b96CA72C696#selector-0x5d8c8844

File: packages/website/src/features/Packages/FunctionInput/TupleInput.tsx (hunk 3 above)
Comment (Ignusmart) [reply packages/website/src/features/Packages/FunctionInput/TupleInput.tsx:11]: Good Catch! sure! Thanks @mjlescano

File: packages/website/src/features/Packages/FunctionInput/TupleInput.tsx (hunk 4 above)
Comment (Ignusmart) [reply packages/website/src/features/Packages/FunctionInput/TupleInput.tsx:12]: Good Catch! sure! Thanks @mjlescano

**Review by noahlitvin — APPROVED** [review]

**Pull request merged**